assets/lambda/ImageOptimization/index.ts

```diff
@@ -0,0 +1,133 @@
+  (bucketName: string) =>
+  async (req: IncomingMessage, res: ServerResponse, url?: NextUrlWithParsedQuery): Promise<void> => {
+    if (!url) {
+      throw new Error('URL is missing from request.')
+    }
+
+    let response: any
+    let data: Buffer
+    // External url, try to fetch image
+    if (url.href?.toLowerCase().startsWith('http')) {
+      try {
+        pipeRes(https.get(url), res)
+      } catch (err) {
+        console.error('Failed to get image', err)
+        res.statusCode = 400
+        res.end()
+      }
+    } else {
+      // S3 expects keys without leading `/`
+      const trimmedKey = url.href.startsWith('/') ? url.href.substring(1) : url.href
+
+      const client = new S3Client({})
+      response = await client.send(new GetObjectCommand({ Bucket: bucketName, Key: trimmedKey }))
+      if (!response.Body) {
+        res.setHeader('Cache-Control', 'no-store,no-cache,must-revalidate,proxy-revalidate')
+        throw new Error(`Could not fetch image ${trimmedKey} from bucket.`)
+      }
+      pipeRes(response.Body, res)
+      // Respect the bucket file's content-type and cache-control
+      // nextImageOptimizer will use this to set the results.maxAge
+      if (response.ContentType) {
+        res.setHeader('Content-Type', response.ContentType)
+      }
+      if (response.CacheControl) {
+        res.setHeader('Cache-Control', response.CacheControl)
+      }
+    }
+  }
+
+// Make header keys lowercase to ensure integrity.
+const normalizeHeaders = (headers: Record<string, any>) =>
+  Object.entries(headers).reduce((acc, [key, value]) => ({ ...acc, [key.toLowerCase()]: value }), {} as Record<string, string>)
+
+const nextConfig: NextConfigComplete = {
+  ...(defaultConfig as NextConfigComplete),
+  images: {
+    ...(defaultConfig.images as ImageConfigComplete),
+    ...config.images
+  },
+}
+
+const optimizer: APIGatewayProxyHandlerV2 = async (event) => {
+  try {
+    if (!sourceBucket) {
+      throw new Error('Bucket name must be defined!')
+    }
+    // This will reject if the image url is not in the acceptable domains
+    // specified in the user's next.config.js config: `domains` and/or `remotePatterns`
+    const imageParams = ImageOptimizerCache.validateParams({ headers: event.headers } as any, event.queryStringParameters!, nextConfig, false)
+    if ('errorMessage' in imageParams) {
+      throw new Error(imageParams.errorMessage)
+    }
+
+    const optimizedResult = await nextImageOptimizer(
+      { headers: normalizeHeaders(event.headers) } as any,
+      {} as any, // res object is not necessary as it's not actually used.
+      imageParams,
+      nextConfig,
+      false, // not in dev mode
+      requestHandler(sourceBucket),
+    )
+
+    return {
+      statusCode: 200,
+      body: optimizedResult.buffer.toString('base64'),
+      isBase64Encoded: true,
+      headers: { 
+        Vary: 'Accept', 
+        'Cache-Control': `public,max-age=${optimizedResult.maxAge},immutable`,
```

Comment:
Should we be using `s-max-age` here and below instead?

Reply:
That's a great question! I think we'd need to give the user the option to control both right? I am not authoritative on this subject

Reply:
I think it should be okay to just use `max-age` for both the browser and CDN to use. Is there a benefit to supplying `x-maxage` w/ a different time than the browser?